File: src/tray/createAppIcon.js
import { app, nativeImage } from 'electron';
import path from 'path';
import plist from 'simple-plist';
import { exec } from 'child_process';
import temp from 'temp';
import fs from 'fs-extra';

import getAppInfo from '../brew/getAppInfo';

export default async function createAppIcon(appInfo) {
    try {
        const appPath = getAppPath(appInfo);
        return getAppIcon(appPath);
    } catch (error) {
        return null;
    }
}


function getAppPath(appInfo) {
    let index = 0;
    const name = appInfo.artifacts.find((artifact) => {
        index = artifact.indexOf('.app');
        if (index === -1) {
            index = artifact.indexOf('.pkg');
        }

        if (index === -1) {
            return false;
        }

        return true;
    });

    return `/Applications/${name.substr(0, index)}.app`;
}


async function getAppIcon(appPath) {
    const info = await readPlist(path.join(appPath, '/Contents/Info.plist'));

    const iconPath = info.CFBundleIconFile.includes('.icns') ? info.CFBundleIconFile : `${info.CFBundleIconFile}.icns`;

    return icnsToNativeImage(path.join(appPath, '/Contents/Resources/', iconPath));
}


const icnsToNativeImage = filePath => new Promise((resolve, reject) => {
    const tempPng = temp.path({ prefix: 'kegger', suffix: '.png' });
    exec(`sips -s format png -Z 16 ${filePath} --out ${tempPng}`, (error) => {
        if (error) {
            return reject(error);
        }

        const icon = resolve(nativeImage.createFromPath(tempPng));
        fs.unlink(tempPng);
        return icon;
    });
});


const readPlist = filePath => new Promise((resolve, reject) => {
    plist.readFile(filePath, (error, result) => {
        if (error) {
            return reject(error);
        }

        return resolve(result);
    });
});

File: internal/tray/CreateAppIcon.go
package tray

import (
	"errors"
	"os"
	"path/filepath"

	"fyne.io/fyne/v2"
	"github.com/bealearts/kegger/internal/brew"
	"howett.net/plist"
)

func CreateAppIcon(appInfo *brew.AppInfo) (fyne.Resource, error) {
	appPath, err := getAppPath(appInfo)
	if err != nil {
		return nil, err
	}

	return getAppIcon(appPath)
}

func getAppPath(appInfo *brew.AppInfo) (string, error) {
	for _, artifact := range appInfo.Artifacts {
		if len(artifact.App) != 0 {
			return "/Applications/" + artifact.App[0], nil
		}
		if len(artifact.Pkg) != 0 {
			fileName := artifact.Pkg[0]
			return "/Applications/" + fileName[:len(fileName)-len(filepath.Ext(fileName))] + ".app", nil
		}
	}

	return "", errors.New("application path not found")
}

func getAppIcon(appPath string) (fyne.Resource, error) {
	iconPath, err := readPListCFBundleIconFile(appPath)
	if err != nil {
		return nil, err
	}

	icon, err := fyne.LoadResourceFromPath(filepath.Join(appPath, iconPath))
	if err != nil {
		return nil, err
	}
	return icon, nil
}

func readPListCFBundleIconFile(appPath string) (string, error) {
	file, err := os.Open(appPath + "/Contents/Info.plist")
	if err != nil {
		return "", err
	}
	defer file.Close()

	type sparseBundle struct {
		CFBundleIconFile string `plist:"CFBundleIconFile"`
	}

	decoder := plist.NewDecoder(file)
	var data sparseBundle
	err = decoder.Decode(&data)
	if err != nil {
		return "", err
	}
	return "/Contents/Resources/" + data.CFBundleIconFile, err
}
